File: ios/Gorilla Groove/BackgroundTaskService.swift
import Foundation

class BackgroundTaskService {
    private static var idToTask: [Int: BackgroundTaskItem] = [:]
    
    private static var polling = false
    
    private static var lock = NSObject()
    
    static var tasks: [BackgroundTaskItem] {
        get {
            return Array(idToTask.values)
        }
    }
    
    private static weak var addTrackListController: AddTrackListController?
    static func registerAddTrackController(_ vc: AddTrackListController) {
        self.addTrackListController = vc
    }
    
    static func addBackgroundTasks(_ tasks: [BackgroundTaskItem]) {
        GGLog.info("Adding new background tasks with IDs \(tasks.map { $0.id })")
        synchronized(lock) {
            tasks.forEach { task in
                idToTask[task.id] = task
            }
            
            addTrackListController?.recalculateProgressText()
            
            if polling {
                GGLog.debug("Already polling for tasks. Not starting a new poll")
                return
            } else {
                GGLog.debug("Not already polling for tasks. Starting a new poll")
                polling = true
                pollUntilAllProcessed()
            }
        }
    }
    
    static func pollUntilAllProcessed() {
        let ids = idToTask.values
            .filter { $0.status == .PENDING || $0.status == .RUNNING }
            .map { $0.id.toString() }
            .joined(separator: ",")
        
        GGLog.debug("Polling for tasks with IDs \(ids)")

        HttpRequester.get("background-task?ids=\(ids)", BackgroundTaskResponse.self) { response, status, _ in
            guard let items = response?.items, status.isSuccessful() else {
                GGLog.error("Could not find background task statuses")
                return
            }
            
            items.forEach { item in
                idToTask[item.id] = item

                if item.status == .FAILED {
                    DispatchQueue.main.async {
                        GGLog.error("Failed to download '\(item.description)'")
                        Toast.show("Failed to download '\(item.description)'")
                    }
                    idToTask[item.id] = nil
                }
            }
            
            synchronized(lock) {
                let allDone = idToTask.values.allSatisfy { $0.status == .COMPLETE }
                GGLog.debug("Polling is \(allDone ? "done" : "not done")")
                if allDone {
                    let completedTasks = idToTask.values
                    
                    ServerSynchronizer.syncWithServer(syncTypes: [.track, .reviewSource], abortIfRecentlySynced: false)
                    
                    DispatchQueue.main.async {
                        if completedTasks.count == 1 {
                            Toast.show("Finished downloading '\(completedTasks.first!.description)'")
                        } else if completedTasks.count > 1 {
                            Toast.show("Finished downloading \(completedTasks.count) items")
                        }
                    }
                    
                    idToTask = [:]
                    polling = false
                } else {
                    DispatchQueue.global().asyncAfter(deadline: .now() + 20.0) {
                        pollUntilAllProcessed()
                    }
                }
                
                addTrackListController?.recalculateProgressText()
            }
        }
    }
}

struct BackgroundTaskItem : Codable {
    let id: Int
    var status: BackgroundProcessStatus
    let type: BackgroundProcessType
    let description: String
}

enum BackgroundProcessStatus : String, Codable {
    case PENDING
    case RUNNING
    case COMPLETE
    case FAILED
}

enum BackgroundProcessType : String, Codable {
    case YT_DOWNLOAD
    case NAMED_IMPORT
}

struct BackgroundTaskResponse : Codable {
    let items: [BackgroundTaskItem]
}

File: android/app/src/main/java/com/gorilla/gorillagroove/service/BackgroundTaskService.kt
package com.gorilla.gorillagroove.service

import com.gorilla.gorillagroove.di.Network
import com.gorilla.gorillagroove.service.GGLog.logDebug
import com.gorilla.gorillagroove.service.GGLog.logError
import com.gorilla.gorillagroove.service.GGLog.logInfo
import com.gorilla.gorillagroove.service.sync.ServerSynchronizer
import com.gorilla.gorillagroove.service.sync.SyncType
import com.gorilla.gorillagroove.util.GGToast
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow

object BackgroundTaskService {
    private var idToTask = mutableMapOf<Long, BackgroundTaskItem>()

    private var polling = false

    val currentDownload = MutableStateFlow(0)
    val totalDownloads = MutableStateFlow(0)

    @Synchronized
    fun addBackgroundTasks(tasks: List<BackgroundTaskItem>) {
        logInfo("Adding new background tasks with IDs ${tasks.map { it.id }}")
        tasks.forEach { task ->
            idToTask[task.id] = task
        }

        totalDownloads.value = totalDownloads.value + tasks.size

        if (polling) {
            logDebug("Already polling for tasks. Not starting a new poll")
            return
        } else {
            logDebug("Not already polling for tasks. Starting a new poll")
            polling = true
            GlobalScope.launch(Dispatchers.IO) {
                pollUntilAllProcessed()
            }
        }
    }

    fun pollIfNeeded() {
        if (idToTask.isNotEmpty() && !polling) {
            GlobalScope.launch(Dispatchers.IO) {
                pollUntilAllProcessed()
            }
        }
    }

    private suspend fun pollUntilAllProcessed() {
        val ids = idToTask.values
            .filter { it.status == BackgroundProcessStatus.PENDING || it.status == BackgroundProcessStatus.RUNNING }
            .map { it.id }
            .joinToString(",")

        logDebug("Polling for tasks with IDs $ids")

        val tasks = try {
            Network.api.getActiveBackgroundTasks(ids)
        } catch (e: Throwable) {
            // This will eventually be restarted by pollIfNeeded() being invoked. Probs not ideal but whatevs.
            logError("Could not get background tasks!", e)
            polling = false

            return
        }

        tasks.items.forEach { task ->
            if (task.status == BackgroundProcessStatus.FAILED) {
                logError("Failed to download '${task.description}'")
                GGToast.show("Failed to download '${task.description}'")

                idToTask.remove(task.id)
            } else {
                idToTask[task.id] = task
            }
        }

        var allDone = false
        var completedTasks: MutableCollection<BackgroundTaskItem> = mutableListOf()

        synchronized(this) {
            allDone = idToTask.values.all { it.status == BackgroundProcessStatus.COMPLETE }

            logDebug("Polling is ${if (allDone) "done" else "not done"}")
            if (allDone) {
                completedTasks = idToTask.values

                idToTask = mutableMapOf()
                polling = false
            }
        }

        currentDownload.value = idToTask.filter { it.value.status == BackgroundProcessStatus.COMPLETE }.size
        totalDownloads.value = idToTask.size

        if (allDone) {
            ServerSynchronizer.syncWithServer(setOf(SyncType.TRACK, SyncType.REVIEW_SOURCE), abortIfRecentlySynced = false)

            if (completedTasks.size == 1) {
                GGToast.show("Finished downloading '${completedTasks.first().description}'")
            } else if (completedTasks.size > 1) {
                GGToast.show("Finished downloading ${completedTasks.size} items")
            }
        } else {
            coroutineScope {
                launch {
                    delay(20_000)
                    pollUntilAllProcessed()
                }
            }
        }
    }
}

data class BackgroundTaskItem(
    val id: Long,
    val status: BackgroundProcessStatus,
    val type: BackgroundProcessType,
    val description: String,
)

enum class BackgroundProcessStatus {
    PENDING,
    RUNNING,
    COMPLETE,
    FAILED,
}

enum class BackgroundProcessType {
    YT_DOWNLOAD,
    NAMED_IMPORT,
}

data class BackgroundTaskResponse(val items: List<BackgroundTaskItem>)
